README.md
WebChess Unofficial 1.0.0rc3 (2013-12-08)
 + changes:
   - Updated the chessdb.php::saveGame() to be more efficient so moves are processed quicker.
   - Added Deutsch translations for gettext. This is disabled in lang.php by default.
 + minor fixes:
   - Fixed issue where you couldn't create a game
   - Fixed issue where you couldn't accept a game
   - Fixed issue where the board wouldn't load
   - Removed PHP short tags
   - Added missing '' ENUM value `games` table columns
   - Added missing `players``userlevel` column
   - Updated `players` table definition to hold the md5 password and change the CHARs to VARCHARs
   - Removed some PHP warnings and notices
   - These fixes were minimal just to get a game up and running and see if it's worth playing
   - Changed the window.onload to document.onready (using domready.js)
   - Fixed issue with columns collapsing if they contained no pieces
   - The replay button titles and text can now be translated
   - Email messages can now be translated
   - Fixed issue where emails weren't being sent (From email address had mistype in variable name)
   - Moved footer html in to footer.php
   - Changes the charset of main pages to UTF-8
   - Replaces the hardcoded app name 'WebChess' with a constant in config.php
   
WebChess 1.0.0rc2 (2010-08-14)
 + minor changes:
   - user interface appearance changes
   - watch also other players' games
   - md5 hash passwords in database

A PHP Web Application that you can install on your own web server. It allows you to play chess with other users across the internet or sitting at the same screen. It only permits valid moves and can automatically detect check and checkmate status.

Original website: http://sourceforge.net/projects/webchess/


docs/CHANGELOG.txt
2013-12-08 (unofficial 1.0.0rc3 version - https://github.com/gitjake/webchess )

- Updated the chessdb.php::saveGame() to be more efficient so moves are processed quicker.
- Added Deutsch translations for gettext. This is disabled in lang.php by default.
- The replay button titles and text can now be translated
- Email messages can now be translated
- Fixed issue where you couldn't create a game
- Fixed issue where you couldn't accept a game
- Fixed issue where the board wouldn't load
- Fixed XSS issue in messages
- Fixed issue where emails weren't being sent (From email address had mistype in variable name)
- Removed PHP short tags
- Added missing '' ENUM value games table columns
- Added missing playersuserlevel column
- Updated players table definition to hold the md5 password and change the CHARs to VARCHARs
- Removed some PHP warnings and notices
- These fixes were minimal just to get a game up and running and see if it's worth playing
- Changed the window.onload to document.onready (using domready.js)
- Fixed issue with columns collapsing if they contained no pieces
- Moved footer html in to footer.php
- Changes the charset of main pages to UTF-8
- Replaces the hardcoded app name 'WebChess' with a constant in config.php

2013-01-10 (unofficial version - https://github.com/Thorium/webchess )
- user interface appearance changes
- watch also other players' games
- user passwords md5 hashed

2010-08-14 (CVS)
- upgraded WebChess license to GPL v3


2004-09-02 (v0.9.0):
- display of captured pieces
- initial stylesheet formatting
- PHP export
- new config setting for board size
- new config setting to restrict the creation of new users
- many, many bug fixes, particularly with check and checkmate detection!!!


2003-10-20 (v0.8.4):
- BUG FIX: fixed a bug in isCheckMate.js where array boundaries weren't being enforced and sometimes caused an invalid index javascript error
- BUG FIX: Checkmate checking fix - by removing king from board isCheckMate() incorrectly believed pieces on the other side of the king could kill the attacking piece, so now king is replaced with ennemy pawn - Bug 753704
- BUG FIX: Fixed bug in validation.js where moving the king along the same line as the attached was not properly validating the move becaues the king was in the way - Bug 750083


2003-05-29 (v0.8.3):
- IMPORTANT: changed name of 'check' column in history table to 'isInCheck' since 'check' is a reserved word in MySQL 4.0
- BUG FIX: fixed case 7 in getNextAttacker() as it was identical to case 6
- BUG FIX: fixed isSafe(): after determining that an attacking king wasn't really a threat because it would be moving into check it would ignore all other possible attackers, which would result in false results under certain circumstances
- BUG FIX: isSafe() sometimes didn't check all the way accross the board (changed i<7 to i<8)
- BUG FIX: fixed a bug in the verbous history settings where pawn promotions wouldn't display what it was promoted to
- BUG FIX: fixed bug whereby check and checkmate weren't properly being detected after a pawn promotion
- BUG FIX: fixed bug where code to update history and theme preferences where inversed (contributed by orcadesign)
- BUG FIX: fixed bug where nick could not be updated (contributed by jdexter)
- IMPORTANT: changed players table to make nick column UNIQUE to prevent users from reusing an existing nick (contributed by Ray Bohac)
- BUG FIX: added a couple sanity checks on the PHP side for empty nicks, but there's still no way to prevent whitespace-only nicks (ToDo)


2002-09-22 (v0.8.2):
- BUG FIX: added 'var' keyword to every local javascript variables to actually make them local
- BUG FIX: allow only one request from a given user A to a given user B (solves problem of mutliplying requests when hiting reload in browser)
- BUG FIX: modified isSafe() to take into account moving the king into check (ToDo: moving into check in general) 
- split list of "current games" into current (ie: active) games and ended games
- added ability to withdraw a pending invitation to play a game
- added "Your Move" or "Opponent's Move" to title and status window
- added ability to edit personal preferences; whether the nick can be changed depends on a server setting since the nick is really how players identify themselves with each other (internally, WebChess uses the player ID)


2002-09-16 (v0.8.2a):
- IMPORTANT: changed preferences table to allow NULL values and values as big as 50 characters (for emails, requires updating the database)
- IMPORTANT: changed games table to add checkMate status (requires updating the database)
- added code to check for missing preferences and saves default preferences for them (makes adding new preferences easier)
- added optional email notification: if a user supplies an email address, they will be notified of opponent's moves
- moved the code to generate the actual PGN/Verbose history string for a given move into chessutils functions (to share with email notification); decided to use the exact same interface for both functions; (could be useful later on: object-oriented WebChess? :)
- when creating new player, added code to verify whether a Nick is already in use
- added second password field to confirm password
- added basic form validation when creating new user
- added logout button to main game screen
- added random color selection when inviting users to a game
- added fix for PHP version < 4.1.0 (http vars used to be HTTP_POST_VARS for example, while WebChess uses _POST)
    NOTE: developers working on WebChess should now make _POST, _GET and _SESSION global in any function they are used, but only for old PHP versions (see chessdb.php for an example)
- BUG FIX: after player A moved his pawn to the last row, but before he did the promotion, player B used to be able to move (credit: Michael Evraire)
- BUG FIX: if a player attempts an illegal move then resigns, the illegal move used to be written to the database (credit: Jay B Chapman)



chess.php
<?php
// $Id: chess.php,v 1.15 2013/12/08 14:00:00 gitjake Exp $

/*
    This file is part of WebChess. http://webchess.sourceforge.net
	Copyright 2010 Jonathan Evraire, Rodrigo Flores

    WebChess is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    WebChess is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with WebChess.  If not, see <http://www.gnu.org/licenses/>.
*/

	session_start();

	/* load settings */
	if (!isset($_CONFIG)) {
		require 'config.php';
		include_once 'lang.php';
	}

	/* define constants */
	require 'chessconstants.php';

	/* include outside functions */
	if (!isset($_CHESSUTILS))
		require 'chessutils.php';
	require 'gui.php';
	require 'chessdb.php';
	require 'move.php';
	require 'undo.php';

	/* allow WebChess to be run on PHP systems < 4.1.0, using old http vars */
	fixOldPHPVersions();

	/* check session status */
	require 'sessioncheck.php';

	/* check if loading game */
	if (isset($_POST['gameID']))
		$_SESSION['gameID'] = $_POST['gameID'];

	/* debug flag */
	define ("DEBUG", 0);

	/* connect to database */
	require 'connectdb.php';

	/* get White's nick */
	$tmpNick = mysql_query("SELECT nick FROM " . $CFG_TABLE[players] . ", " . $CFG_TABLE[games] . " WHERE playerID = whitePlayer AND gameID = " . $_SESSION['gameID']);
	$whiteNick = mysql_result($tmpNick, 0);

	/* get Black's nick */
	$tmpNick = mysql_query("SELECT nick FROM " . $CFG_TABLE[players] . ", " . $CFG_TABLE[games] . " WHERE playerID = blackPlayer AND gameID = " . $_SESSION['gameID']);
	$blackNick = mysql_result($tmpNick, 0);

	/* load game */
	$isInCheck = (isset($_POST['isInCheck']) && ($_POST['isInCheck'] == 'true'));
	$isCheckMate = false;
	$isPromoting = false;
	$isUndoing = false;
	loadHistory();
	loadGame();
	processMessages();

	if ($isUndoing)
	{
		doUndo();
		saveGame();
	}
	elseif (!empty($_POST['promotion']) && isset($_POST['toRow']) && ('' !== $_POST['toRow']) && isset($_POST['toCol']) && ('' !== $_POST['toCol'])) 
	{
		savePromotion();
		$board[$_POST['toRow']][$_POST['toCol']] = $_POST['promotion'] | ($board[$_POST['toRow']][$_POST['toCol']] & BLACK);
		saveGame();
	} 
	elseif (
		isset($_POST['fromRow']) && isset($_POST['fromCol']) && isset($_POST['toRow']) && isset($_POST['toCol'])
		&& ('' !== $_POST['fromRow']) && ('' !== $_POST['fromCol']) && ('' !== $_POST['toRow']) && ('' !== $_POST['toCol'])
	) // END elseif
	{
		/* ensure it's the current player moving				 */
		/* NOTE: if not, this will currently ignore the command...               */
		/*       perhaps the status should be instead?                           */
		/*       (Could be confusing to player if they double-click or something */
		$tmpIsValid = true;
		if (($numMoves == -1) || ($numMoves % 2 == 1))
		{
			/* White's move... ensure that piece being moved is white */
			if ((($board[$_POST['fromRow']][$_POST['fromCol']] & BLACK) != 0) || ($board[$_POST['fromRow']][$_POST['fromCol']] == 0))
				/* invalid move */
				$tmpIsValid = false;
		}
		else
		{
			/* Black's move... ensure that piece being moved is black */
			if ((($board[$_POST['fromRow']][$_POST['fromCol']] & BLACK) != BLACK) || ($board[$_POST['fromRow']][$_POST['fromCol']] == 0))
				/* invalid move */
				$tmpIsValid = false;
		}

		if ($tmpIsValid)
		{
			saveHistory();
			doMove();
			saveGame();
		}
	}
	elseif($history[$numMoves]['curPiece'] == 'pawn' && $history[$numMoves]['promotedTo'] == null)
	{	// Incomplete promotion?
		if($history[$numMoves]['toRow'] == 7 || $history[$numMoves]['toRow'] == 0)
		{
			$isPromoting = true;
		}
	}

	mysql_close();
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
   "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="pragma" content="no-cache" />
<meta http-equiv="Page-Exit" content="blendTrans(Duration=0.18)">
<link rel="stylesheet" href="chess.css" type="text/css" />
<?php
	echo("<link rel='stylesheet' href='images/");
	echo($_SESSION['pref_theme'] . "/wctheme.css' type='text/css' />\n");

	/* find out if it's the current player's turn */
	if (( (($numMoves == -1) || (($numMoves % 2) == 1)) && ($playersColor == "white"))
			|| ((($numMoves % 2) == 0) && ($playersColor == "black")))
		$isPlayersTurn = true;
	else
		$isPlayersTurn = false;

	if ($_SESSION['isSharedPC'])
		echo '<title>', APP_NAME, "</title>\n";
	else if ($isPlayersTurn)
		echo '<title>', APP_NAME . gettext(' - Your Move'), "</title>\n";
	else
		echo '<title>', APP_NAME . gettext(" - Opponent's Move"), "</title>\n";
?>
<script type="text/javascript">
<?php
	echo("var cfgImageExt = '$CFG_IMAGE_EXT';\n");
	/* transfer board data to javacript */
	writeJSboard();
	/* if it's not the player's turn, enable auto-refresh */
	$autoRefresh = !$isPlayersTurn && !isBoardDisabled() && !$_SESSION['isSharedPC'];
	echo('var autoreload = ');
	if(!$autoRefresh)
		echo('0');
	else if ($_SESSION['pref_autoreload'] >= $CFG_MINAUTORELOAD)
		echo ($_SESSION['pref_autoreload']);
	else
		echo ($CFG_MINAUTORELOAD);
	echo(";\n");
	writeJShistory();
	drawboard();
	echo 'var gameId = ' . $_SESSION['gameID'] . ";\n";
	echo 'var players = "' . $whiteNick . ' - ' . $blackNick . "\";\n";
	echo 'var playersColor = "' . $playersColor . "\";\n";
	echo 'var isPromoting = "'.$isPromoting. "\";\n";
	echo 'var isKingInCheck = "'.$isInCheck. "\";\n";
	echo 'var isGameOver = "'.$isGameOver. "\";\n";
	echo 'var historyLayout = "'.$_SESSION['pref_historylayout']. "\";\n";
?>
	// I18n Messages. The translation lang is set in I18N_LOCALE and requires you to have the matching webchess.mo 
	var i18nMessages = {
		"Start of game": "<?php echo gettext('Start of game'); ?>"
		,"Start": "<?php echo gettext('Start'); ?>"
		,"Go back five halfmoves": "<?php echo gettext('Go back five halfmoves'); ?>"
		,"Go back one halfmove": "<?php echo gettext('Go back one halfmove'); ?>"
		,"Go forward one halfmove": "<?php echo gettext('Go forward one halfmove'); ?>"
		,"Go forward five halfmoves": "<?php echo gettext('Go forward five halfmoves'); ?>"
		,"End of game": "<?php echo gettext('End of game'); ?>"
		,"End": "<?php echo gettext('End'); ?>"
	};
<?php
	
	writeStatus();
	writeHistory();
	// Captured pieces..
	require 'capt.php';
?>
</script>
<script type="text/javascript" src="javascript/domready.js"></script>
<script type="text/javascript" src="javascript/chessutils.js"></script>
<script type="text/javascript" src="javascript/commands.js"></script>
<script type="text/javascript" src="javascript/validation.js"></script>
<?php
if($isPlayersTurn || $_SESSION['isSharedPC'] || $isPromoting)
	echo('
<script type="text/javascript" src="javascript/isCheckMate.js"></script>');
if(!isBoardDisabled() || $_SESSION['isSharedPC'])
	echo('
<script type="text/javascript" src="javascript/squareclicked.js"></script>');
?>
<script type="text/javascript" src="javascript/board.js"></script>
</head>
<body>
<div id="wrapper">
	<div id="header">
	  <div id="heading"><?php echo APP_NAME; ?></div>
	</div>
    <?php require 'info.php'; ?>
	<div id="boardsection" align="center">
		<form name="gamedata" method="post" action="chess.php">
		<?php
			if ($isPromoting && (!$isPlayersTurn || $_SESSION['isSharedPC'])) // Write promotion dialog only to the correct player
				writePromotion();
			if ($isUndoRequested)
				writeUndoRequest();
			if ($isDrawRequested)
				writeDrawRequest();
		?>
		<div id="chessboard"></div>
		<div id="moveinfo">
			<div id="curmove">&nbsp;</div>
			<div id="whosmove">&nbsp;</div>
		</div>
		<div id="gamebuttons">
		<input type="button" id="btnUndo" class="button" value="Request Undo" disabled="disabled" />
		<input type="button" id="btnDraw" class="button" value="Request Draw" disabled="disabled" />
		<input type="button" id="btnResign" class="button" value="Resign" disabled="disabled" />
		</div>
		<input type="hidden" name="requestUndo" value="no" />
		<input type="hidden" name="requestDraw" value="no" />
		<input type="hidden" name="resign" value="no" />
		<input type="hidden" name="fromRow" value="<?php if ($isPromoting) echo ($_POST['fromRow']); ?>" />
		<input type="hidden" name="fromCol" value="<?php if ($isPromoting) echo ($_POST['fromCol']); ?>" />
		<input type="hidden" name="toRow" value="<?php if ($isPromoting) echo ($_POST['toRow']); ?>" />
		<input type="hidden" name="toCol" value="<?php if ($isPromoting) echo ($_POST['toCol']); ?>" />
		<input type="hidden" name="isInCheck" value="false" />
		<input type="hidden" name="isCheckMate" value="false" />
		</form>
		<div id="gamenav"></div>
		<div><?echo gettext('When castling, just move the king (the rook will move automatically).');?></div>
		<div id="captheading"><?echo gettext('Captured pieces'); ?></div>
		<div id="captures"></div>
	</div>

	<div id="content">
		<div id="players" class="move_header"></div>
		<div id="gameid" class="move_header"></div>
		<div id="gamebody" class="move_header"></div>
		<div id="checkmsg"></div>
		<div id="statusmsg"></div>
		<form name="gamemenu" method="post" action="chess.php" style="text-align:center;">
		<input type="button" id="btnMainMenu" class="button" value="Menu" disabled="disabled" />
		<input type="button" id="btnReload" class="button" value="Reload" disabled="disabled" />
		<input type="button" id="btnPGN" class="button" value="PGN" disabled="disabled" />
		<input type="button" id="btnLogout" class="button" value="Logout" disabled="disabled" />
		<input type="hidden" name="ToDo" value="Logout" />	<!-- NOTE: this field is only used to Logout -->
		</form>
	</div>
	<?php include_once('footer.php'); ?>
</div>
</body>
</html>
